client/canonical_eventlog.go

```diff
@@ -0,0 +1,266 @@
+package client
+
+import (
+	"bytes"
+	"crypto"
+	"encoding/binary"
+	"fmt"
+	"io"
+
+	"github.com/google/go-tpm/tpm2"
+)
+
+var hashAlgoMapping = map[crypto.Hash]tpm2.Algorithm{
+	crypto.SHA1:   tpm2.AlgSHA1,
+	crypto.SHA256: tpm2.AlgSHA256,
+	crypto.SHA384: tpm2.AlgSHA384,
+	crypto.SHA512: tpm2.AlgSHA512,
+}
+
+const (
+	// CEL spec 5.1
+	recNumTagValue uint8 = 0
+	pcrTagValue    uint8 = 1
+	// nvIndexTagValue uint8 = 2 // nvindex field is not supported
+	digestsTagValue uint8 = 3
+
+	tlvTagFieldLength    int = 1
+	tlvLengthFieldLength int = 4
+
+	recNumValueLength uint32 = 8 // support up to 2^64 records
+	pcrValueLength    uint32 = 1 // support up to 256 PCRs
+)
+
+// TLV definition according to CEL spec, page 16
+type TLV struct {
+	Tag    uint8
+	Length uint32 // big-endian when encoding
+	Value  []byte // with size of Length
+}
+
+// TLVMarshal marhsal a TLV to a byte slice
+func TLVMarshal(tlv TLV) []byte {
```

Comment:
I don't see a strong reason to pass a pointer or value here, but I changed it to pointer

Reply:
a value receiver *copies* the type on calling the function; this makes it undesirable on a TLV-like struct where it can be arbitrarily big:
See https://github.com/golang/go/wiki/CodeReviewComments#receiver-type

Reply:
also, sorry I was unclear: let's change the name to just `Marshal`

so: `func (tlv *TLV) Marshal() ([]byte, error) {`
The TLV is implied

Reply:
Done